Framework/Crystal/src/SCDCalibratePanels.cpp

```diff
@@ -605,677 +152,374 @@ void SCDCalibratePanels::exec() {
+  OrientedLattice lattice = peaksWs->mutableSample().getOrientedLattice();
+  DblMatrix UB = lattice.getUB();
+  // sort again since edge peaks can trace to other banks
+  peaksWs->sort(criteria);
+  PARALLEL_FOR3(ColWksp, RowWksp, TofWksp)
+  for (int i = 0; i < static_cast<int>(MyBankNames.size()); ++i) {
+    PARALLEL_START_INTERUPT_REGION
+    std::set<string>::iterator it = MyBankNames.begin();
+    advance(it, i);
+    std::string bankName = *it;
+    size_t k = bankName.find_last_not_of("0123456789");
+    int bank = 0;
+    if (k < bankName.length())
+      bank = boost::lexical_cast<int>(bankName.substr(k + 1));
+    ColWksp->getSpectrum(i).setSpectrumNo(specnum_t(bank));
+    RowWksp->getSpectrum(i).setSpectrumNo(specnum_t(bank));
+    TofWksp->getSpectrum(i).setSpectrumNo(specnum_t(bank));
+    int icount = 0;
+    for (int j = 0; j < nPeaks; j++) {
+      Peak peak = peaksWs->getPeak(j);
+      if (peak.getBankName() == bankName) {
+        try {
+          V3D q_lab =
+              (peak.getGoniometerMatrix() * UB) * peak.getHKL() * M_2_PI;
+          Peak theoretical(peak.getInstrument(), q_lab);
+          ColWksp->dataX(i)[icount] = peak.getCol();
+          ColWksp->dataY(i)[icount] = theoretical.getCol();
+          RowWksp->dataX(i)[icount] = peak.getRow();
+          RowWksp->dataY(i)[icount] = theoretical.getRow();
+          TofWksp->dataX(i)[icount] = peak.getTOF();
+          TofWksp->dataY(i)[icount] = theoretical.getTOF();
+        } catch (...) {
+          // g_log.debug() << "Problem only in printing peaks\n";
+        }
+        icount++;
+      }
+    }
+    PARALLEL_END_INTERUPT_REGION
   }
-  if (cell_type == ReducedCell::TRICLINIC() ||
-      cell_type == ReducedCell::MONOCLINIC()) {
-    fn->setParameter("Beta", beta);
+  PARALLEL_CHECK_INTERUPT_REGION
+
+  string colFilename = getProperty("ColFilename");
+  string rowFilename = getProperty("RowFilename");
+  string tofFilename = getProperty("TofFilename");
+  saveNexus(colFilename, ColWksp);
+  saveNexus(rowFilename, RowWksp);
+  saveNexus(tofFilename, TofWksp);
+}
+
+void SCDCalibratePanels::saveNexus(std::string outputFile,
+                                   MatrixWorkspace_sptr outputWS) {
+  IAlgorithm_sptr save = this->createChildAlgorithm("SaveNexus");
+  save->setProperty("InputWorkspace", outputWS);
+  save->setProperty("FileName", outputFile);
+  save->execute();
+}
+
+void SCDCalibratePanels::findL1(int nPeaks,
+                                DataObjects::PeaksWorkspace_sptr peaksWs) {
+  MatrixWorkspace_sptr L1WS = boost::dynamic_pointer_cast<MatrixWorkspace>(
+      API::WorkspaceFactory::Instance().create("Workspace2D", 1, 3 * nPeaks,
+                                               3 * nPeaks));
+
+  auto &outSp = L1WS->getSpectrum(0);
+  MantidVec &yVec = outSp.dataY();
+  MantidVec &eVec = outSp.dataE();
+  MantidVec &xVec = outSp.dataX();
+
+  for (int i = 0; i < nPeaks; i++) {
+    DataObjects::Peak peak = peaksWs->getPeak(i);
```

Comment:
Does one need a copy or is the const reference returned by getPeak(...) sufficient?